libraries/Microsoft.Bot.Builder.Ai/TranslationMiddleware.cs

```diff
@@ -0,0 +1,177 @@
+                        var sourceLanguage = "";
+                        if (_getUserLanguage == null)
+                            sourceLanguage = await Task.Run(() => _translator.Detect(message.Text)); // context.Conversation.Data["Language"]?.ToString() ?? this._nativeLanguages.FirstOrDefault() ?? "en";
+                        else
+                        {
+                            sourceLanguage = _getUserLanguage(context);
+                        } 
+                        if (_patterns.ContainsKey(sourceLanguage) && _patterns[sourceLanguage].Count>0)
+                        {
+                            this._translator.SetPostProcessorTemplate(_patterns[sourceLanguage]);
+                        }
+                        if (!_nativeLanguages.Contains(sourceLanguage))
+                        {
+                            var translationContext = new TranslationContext
+                            {
+                                SourceText = message.Text,
+                                SourceLanguage = sourceLanguage,
+                                TargetLanguage = (this._nativeLanguages.Contains(sourceLanguage)) ? sourceLanguage : this._nativeLanguages.FirstOrDefault() ?? "en"
+                            };
+
+                            // translate to bots language
+                            if (translationContext.SourceLanguage != translationContext.TargetLanguage)
+                                await TranslateMessageAsync(context, message, translationContext.SourceLanguage, translationContext.TargetLanguage).ConfigureAwait(false);
+                        }
+                    }
+                }
+            }
+            await next().ConfigureAwait(false);
+        }
+
+
+        private static readonly Regex UrlRegex = new Regex(@"(https?://[^\s]*)", RegexOptions.Compiled | RegexOptions.IgnoreCase | RegexOptions.Multiline);
+
+        /// <summary>
+        /// Translate .Text field of a message
+        /// </summary>
+        /// <param name="context"/>
+        /// <param name="message"></param>
+        /// <param name="sourceLanguage"/>
+        /// <param name="targetLanguage"></param>
+        /// <returns></returns>
+        private async Task TranslateMessageAsync(ITurnContext context, IMessageActivity message, string sourceLanguage, string targetLanguage)
+        {
+            // if we have text and a target language
+            if (!String.IsNullOrWhiteSpace(message.Text) && !String.IsNullOrEmpty(targetLanguage))
+            {
+                if (targetLanguage == sourceLanguage)
+                    return;
+
+                var text = message.Text.Length <= 65536 ? message.Text : message.Text.Substring(0, 65536);
+
+                // massage mentions and urls so they don't get translated
+                int i = 0;
+
+                var urls = new List<string>();
+                while (UrlRegex.IsMatch(text))
```

Comment:
I am 100% confident I could make this code behave incorrectly by throwing unicode URLs at it. The Regex (which is wrong) is also only testing for 'https', not any of the other schemes ('http', for example).

Reply:
this part is removed already handled internally in the Microsoft Translator API